Linking: perform macro expansion and type updates just once

tautschnig opened the pull request:
This reduces the Linux kernel link time by more than 90% as the previous
approach would repeatedly attempt renamings and replacements across all
hitherto read goto program instructions.

Depends on:
- [x] #1260

tautschnig committed:
Encapsulate reading and linking multiple files

We do the same work in multiple places, and can optimize the linking process
when knowing about all files to be linked in one step.

tautschnig committed:
Linking: perform macro expansion and type updates just once

This reduces the Linux kernel link time by more than 90% as the previous
approach would repeatedly attempt renamings and replacements across all
hitherto read goto program instructions.

smowton commented on src/goto-cc/compile.cpp:
```diff
@@ -348,12 +348,19 @@ bool compilet::link()
-  }
+  goto_modelt goto_model;
+
+  goto_model.symbol_table.swap(symbol_table);
```
Surely just introduce a <symbol-table, goto-functions> overload -- or perhaps use `wrapper_goto_modelt`?

tautschnig replied:
Hmm, none of `wrapper_goto_modelt` nor `abstract_goto_modelt` provide non-const accessors. Is there are good reason those should not be added? (I strictly need non-const access here.)

smowton replied:
`abstract_goto_modelt` should not provide a non-const accessor (as lazy-goto-model must enforce that you only use its proper interface to populate new functions, for example). However `wrapper_goto_modelt` and `goto_modelt` may reasonably implement something stronger with non-const accessors.

smowton commented on src/goto-programs/link_goto_model.cpp:
```diff
@@ -107,6 +106,51 @@ static bool link_functions(
+      weak_symbols.insert(symbol_pair.first);
+  }
+
+  linkingt linking(dest.symbol_table,
```
Style (break after open-paren)

smowton commented on src/goto-programs/initialize_goto_model.cpp:
```diff
@@ -38,9 +38,7 @@ goto_modelt initialize_goto_model(
-  std::vector<std::string> binaries, sources;
-  binaries.reserve(files.size());
-  sources.reserve(files.size());
+  std::list<std::string> binaries, sources;
```
`std::list` is quite surprising -- do we really need any of `list`'s properties (cheap insert / delete, stable iterators)? If not I'd stick with `vector`

tautschnig replied:
I'd prefer to say "use whatever SequenceContainer you want," but then I can't do that unless I add a template parameter. The difference will not be measurable. The number of entries will typically be a very small number, with the 1000-2000 entries that the Linux kernel yields being close to an upper bound. I have picked `std::list`, because the number of entries is not known a priori in `compilet`, and hence append-at-the-end would be inefficient with `std::vector`. Again, "inefficient" in theory, practically not measurably so.

smowton replied:
ok, not a big deal either way

tautschnig replied:
I'm not going to insist on using `std::list` either - but I did have to change one or the other, and if you want me to flip a coin I can do that as well :-)

smowton reviewed (approved):
Minor comments, looks good. Can you think of any corner cases to test aren't broken, e.g. unifying two units with a weak defn vs. one with a strong defn, or similarly two with incomplete definitions of a struct vs. one that provides the definition? If so are you happy we've already got regression tests taking care of those?

tautschnig commented:
Just a note: it seems I cannot further postpone fixing #1490, which is what the CI failures are.

tautschnig commented:
I am now including some commits from #1260.

tautschnig commented:
I am working on further fixes in #1260, which should then get merged before this one.

tautschnig committed:
symtab2gb: link symbol tables, not goto models

There is no need to do the additional work that linking on goto
functions entails when the input is just symbol tables. Thus link symbol
tables, and only after linking all of them generate goto functions and
then write the combined goto model.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e868217ffc777cb7c5787d6eb4a166a4e109579f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113340771

martin-cs commented on src/goto-programs/read_goto_binary.cpp:
```diff
@@ -14,18 +14,207 @@ Module: Read Goto Programs
+
+  goto_programt &program = function.body;
+
+  Forall_goto_program_instructions(iit, program)
```
At the risk of submitting a piss-take comment -- do you really want to add a macro for-all given the other PRs of yours I have open?

tautschnig replied:
I could fix this here, but those lines are just an artefact of moving the code from `link_goto_model.cpp` to `read_goto_binary.cpp`, and I'd hence prefer to keep such changes in a separate commit.

martin-cs replied:
Fair.

martin-cs commented on src/goto-programs/read_goto_binary.cpp:
```diff
@@ -14,18 +14,207 @@ Module: Read Goto Programs
+  namespacet src_ns(src_symbol_table);
+
+  // merge functions
+  Forall_goto_functions(src_it, src_functions)
```
Ditto

martin-cs commented on src/symtab2gb/symtab2gb_parse_options.cpp:
```diff
@@ -81,11 +84,18 @@ static void run_symtab2gb(
-    goto_model.symbol_table = symtab;
-    goto_convert(goto_model, message_handler);
-    link_goto_model(linked_goto_model, goto_model, message_handler);
+
```
Am I right in thinking this is actually a larger change?  As I understand it instead of converting each `symbol_tablet` to a `goto_functionst` and linking the `goto_functionst` it is linking the `symbol_tablet`s and then doing one conversion.  Is this correct?  I think this might be a better way of doing things.

tautschnig replied:
Yes, you are absolutely right. I should have said that this PR should be reviewed commit-by-commit (and I hope the commit message of the first commit 3325c3f actually explains this).

martin-cs replied:
Thanks for the explanation.

TGWDB commented:
@tautschnig This PR looks like it is close being merged, apart from codeowner review, is there anything outstanding here?

martin-cs commented:
Also paging @chrisr-diffblue and @peterschrammel

chrisr-diffblue reviewed (approved):
That looks like a sensible improvement :-)

tautschnig commented:
> @tautschnig This PR looks like it is close being merged, apart from codeowner review, is there anything outstanding here?

I had to do a rebase to resolve conflicts, but now it's really just codeowner reviews that are missing.
